model/src/main/java/org/projectnessie/model/Branch.java

```diff
@@ -57,24 +52,43 @@ static Branch of(String name, String hash) {
   }
 
   /**
-   * Returns a map of metadata properties that describe this branch. Note that these properties
-   * <b>can be added</b> by the server when a {@link Branch} instance is returned. A returned {@link
-   * Branch} instance can then have the following keys:
+   * Returns a {@link BranchMetadata} instance that contains additional metadata about this branch.
+   * Note that this is <b>only added</b> by the server when <b>explicitly</b> requested by the
+   * client.
    *
-   * <ul>
-   *   <li>{@value Branch#NUM_COMMITS_AHEAD}: number of commits ahead of the default branch
-   *   <li>{@value Branch#NUM_COMMITS_BEHIND}: number of commits behind the default branch
-   *   <li>{@value Branch#HEAD_COMMIT_META}: the commit metadata of the HEAD commit
-   *   <li>{@value Branch#COMMON_ANCESTOR_HASH}: the hash of the common ancestor in relation to the
-   *       default branch.
-   * </ul>
-   *
-   * @return A map of metadata properties that describe this branch. Note that these properties
-   *     <b>can be added</b> by the server when a {@link Branch} instance is returned.
+   * @return A {@link BranchMetadata} instance that contains additional metadata about this branch.
+   *     Note that this is <b>only added</b> by the server when <b>explicitly</b> requested by the
+   *     client.
    */
-  @Value.Default
-  @JsonInclude(Include.NON_EMPTY)
-  default Map<String, Object> metadataProperties() {
-    return Collections.emptyMap();
+  @JsonInclude(Include.NON_NULL)
+  @Nullable
+  BranchMetadata branchMetadata();
+
+  @Schema(
+      type = SchemaType.OBJECT,
+      title = "Additional metadata for a 'Branch'.",
+      description =
+          "Only returned by the server when explicitly requested by the client and contains the following information:\n\n"
+              + "- numCommitsAhead (number of commits ahead of the default branch)\n\n"
+              + "- numCommitsBehind (number of commits behind the default branch)\n\n"
+              + "- commitMetaOfHEAD (the commit metadata of the HEAD commit)\n\n"
+              + "- commonAncestorHash (the hash of the common ancestor in relation to the default branch).\n\n")
+  @Value.Immutable
+  @JsonSerialize(as = ImmutableBranchMetadata.class)
+  @JsonDeserialize(as = ImmutableBranchMetadata.class)
+  @JsonTypeName("BRANCH_METADATA")
+  interface BranchMetadata {
```

Comment:
I wonder if it makes sense to make this a top-level interface :thinking: Nessie backend is capable of producing this kind metadata for tags as well.

Reply:
that makes sense. I moved this into `ReferenceMetadata` and will add the metadata properties to Tags in the next PR